gobblin-core/src/test/java/org/apache/gobblin/writer/PartitionedWriterTest.java

```diff
@@ -176,4 +165,31 @@ public void testControlMessageHandler() throws IOException {
 
     writer.close();
   }
+
+  @Test
+  public void testPartitionWriterCacheRemovalListener()
+      throws IOException, InterruptedException {
+    State state = new State();
+    state.setProp(ConfigurationKeys.WRITER_PARTITIONER_CLASS, TestPartitioner.class.getCanonicalName());
+    state.setProp(PartitionedDataWriter.PARTITIONED_WRITER_CACHE_TTL_SECONDS, 1);
+    TestPartitionAwareWriterBuilder builder = new TestPartitionAwareWriterBuilder();
+
+    PartitionedDataWriter writer = new PartitionedDataWriter<String, String>(builder, state);
+
+    String record1 = "abc";
+    writer.writeEnvelope(new RecordEnvelope(record1));
+
+    String record2 = "123";
+    writer.writeEnvelope(new RecordEnvelope(record2));
+
+    //Sleep for more than cache expiration interval
+    Thread.sleep(1500);
```

Comment:
Is there a way to test this without Thread.sleep, e.g. with some virtual time or by explicitly triggering expiration? 
Using sleeps can slow down the test suite, and also make it flakey. For example, if the tests are running on overloaded Tracis CI machine, the cache timer that normally runs every second, can run every 3rd second, and the test will randomly fail.

Reply:
Guava does provide Ticker interface for testing cache evictions without having to put sleeps. Unfortunately, the cache has to be built with the ticker, which is not what we want for our test. In our test, since we explicitly call cache cleanup(), we should not encounter the issue you are describing.